R/fit_peak.R

```diff
@@ -0,0 +1,173 @@
+#'       - For 'weibull': Shape parameter (k).
+#'       - For 'lnorm': Mean of the log-transformed observations.
+#'       - For 'exp': Rate parameter (λ).
+#'   - 'optim_fit_par_2': The second fit parameter for the chosen family.
+#'      - For 'weibull': Scale parameter (λ).
+#'      - For 'lnorm': Standard deviation of the log-transformed observations.
+#'      - For 'exp': Not applicable (set to NA).
+#'   - 'obj_value': The value of the objective function
+#'     (negative log-likelihood), which represent the minimized objective
+#'      function value from the optimisation. Smaller value equals better
+#'      optimisation.
+#'   - 'converged': Logical. TRUE if the optimisation converged.
+#'   - 'family': The distribution family used for the optimization.
+#'      - 'weibull': Uses the Weibull distribution for fitting.
+#'      - 'lnorm': Uses the Log-normal distribution for fitting.
+#'      - 'exp': Uses the Exponential distribution for fitting.
+#'
+#' @export
+#'
+#' @examples
+#' # Create three seasons with random observations
+#' obs = 10
+#' season = c("2018/2019", "2019/2020", "2020/2021")
+#' season_num_rev <- rev(seq(from = 1, to = length(season)))
+#' observations = rep(stats::rnorm(10,obs), length(season))
+#'
+#' # Add into a tibble with weight decreasing going one season step back
+#' peak_input <- tibble::tibble(
+#'   observation = observations,
+#'   weight = 0.8^rep(season_num_rev, each = obs)
+#' )
+#'
+#' # Use the peak model
+#' fit_peak(weighted_observations = peak_input,
+#'          conf_levels = c(0.50, 0.90, 0.95),
+#'          family = "weibull")
+
+fit_peak <- function(
+  weighted_observations,
+  conf_levels = c(0.50, 0.90, 0.95),
+  family = c("weibull",
+             "lnorm",
+             "exp"),
+  optim_method = c("Nelder-Mead",
+                   "BFGS",
+                   "CG",
+                   "L-BFGS-B",
+                   "SANN",
+                   "Brent"),
+  lower_optim = -Inf,
+  upper_optim = Inf
+) {
+  # Check input arguments
+  coll <- checkmate::makeAssertCollection()
+  checkmate::assert_tibble(weighted_observations, add = coll)
+  checkmate::assert_names(colnames(weighted_observations),
+                          identical.to = c("observation", "weight"), add = coll)
+  checkmate::assert_numeric(lower_optim, add = coll)
+  checkmate::assert_numeric(upper_optim, add = coll)
+  checkmate::reportAssertions(coll)
+  # Match the arguements.
+  family <- rlang::arg_match(family)
+  optim_method <- rlang::arg_match(optim_method)
+
+  # Initialising parameters based on family
+  init_par_fun <- function(family, observations) {
+    init_params <- switch(family,
+      weibull = log(c(1.5, mean(observations))),
+      lnorm = c(mean(log(observations)),
+                stats::sd(log(observations))),
+      exp = log(1.5)
+    )
+    return(init_params)
+  }
+
+  # The weighted negative loglikelihood function
+  nll <- function(par, weighted_observations, family = "weibull") {
```

Comment:
```suggestion
  nll <- function(par, weighted_observations, family = family) {
```

Right now, it is always using weibull?